src/order/compactly_generated.lean

```diff
@@ -231,4 +253,48 @@ begin
   exact λ x, ⟨{x}, ⟨λ x _, h x, Sup_singleton⟩⟩,
 end
 
+/-- A compact element k has the property that any b < k lies below a "maximal element below k",
+which is to say [⊥, k] is coatomic. -/
+theorem Iic_coatomic_of_compact_element {k : α} (h : is_compact_element k) :
+  is_coatomic (set.Iic k) :=
+begin
+  refine_struct { .. }, rintros ⟨b, hbk⟩,
+  -- The cases b = k or k = ⊥ are both the trivial case b = ⊤ in Iic k.
+  by_cases htriv : b = k ∨ k = ⊥,
```

Comment:
If anyone has ideas to simplify this proof of the trivial case(s), I'd love to hear it. I felt there was some friction here dealing with the subtypes.

Reply:
Idea 1: separately show that `is_compact_element \top` implies `is_coatomic` and that `is_compact_element k` implies `is_compact_element k` in `set.Iic k`

Reply:
Idea 2: Judging by the example of `ideal.exists_maximal`, I think that all compactly generated lattices are coatomic. In that case, it might be better to show in a compactly generated lattice, `set.Iic k` is always compactly generated, and then derive coatomicity.

Reply:
Idea 3: provide a lemma that `subsingleton` implies `is_coatomic`

Reply:
Copying the proof of `ideal.exists_maximal`, I got this to build:
`⟨λ ⟨b, hbk⟩, begin
  by_cases htriv : b = k,
  { left,
    ext,
    simp [htriv] },
  right,
  rcases zorn.zorn_partial_order₀ (set.Iio k) _ b (lt_of_le_of_ne hbk htriv) with ⟨a, a0, ba, h⟩,
  { refine ⟨⟨a, le_of_lt a0⟩, ⟨ne_of_lt a0, λ c hck, by_contradiction $ λ c0, _⟩, ba⟩,
    cases h c.1 (lt_of_le_of_ne c.2 (λ con, c0 (subtype.ext con))) (le_of_lt hck),
    exact lt_irrefl _ hck },
  { exact λ S SC cC I IS, ⟨Sup S, Sup_lt_of_directed_set_lt_compact _ _ h
      (ne_bot_of_gt (lt_of_le_of_ne hbk htriv)) _ cC.directed_on SC, le_Sup _⟩ }
end⟩`